Fix EAP

jsebechlebsky opened the pull request:
Hi,
I discovered EAP-Identity, EAP-TLS, EAP-FAST and EAP-CHAP are not implemented properly, so I've fixed that. I've changed corresponding classes to be subclasses of EAP instead of separate layers. This makes implementation much easier given that parsing each of them properly requires accesing "len" field of EAP layer.

EAP-Identity was just missing optional string field in request.
Main problem with the EAP-TLS and EAP-FAST was that current implementation was not dealing with fragmented messages which are quite common
 EAP-CHAP was not parsing requests properly, since these can contain random challenge value of arbitrary length and length of extra data must be determined from len field of EAP layer. 

There is some more info about the changes in each of the commit messages.
Implementation in the pull request is following relevant RFCs so it should manage to dissect all valid messages properly.

plorinquer commented:
Hello,

I allow myself a few comments since I made a few changes to the existing EAP implementation.


#### EAP-TLS / EAP-FAST

Indeed, the current code does not use the underlayer `len` field, so the payload in packets without the `L` bit set (typically, the last fragment) is not decoded as `tls_data` (and remains in the payload).

Actually, I was waiting for TLS support to be added to update the code, but I did not work on it since it was merged last month. I think subclassing EAP makes things simpler. Thanks for your improvements !


#### EAP-MD5

Indeed, the challenge can be something else than the output of the MD5 hashing function, so one should not assume a fixed length of 16 bytes.


Section 4 of the [RFC3748](https://tools.ietf.org/html/rfc3748#section-4) states that the _Length_ field indicates the length of the EAP packet including the _Code_, _Identifier_, _Length_, and _Data_ fields. In your last commit, the `optional_name` field length should be computed as follows:

`StrLenField("optional_name", '', length_from=lambda p: p.len - p.value_size - 6)`

instead of 

`StrLenField("optional_name", '', length_from=lambda p: p.len - p.value_size - 1)`


One remark regarding the tests in `regression.uts` . In the tests _EAP - EAP_MD5 - Request - Dissection (8)_ and _EAP - EAP_MD5 - Response - Dissection (9)_, the EAP packet length is 22 bytes.

Code : 1 byte
Identifier : 1 byte
Length : 2 bytes
Type : 1 byte
Value-Size: 1 byte
Value: 16 bytes (in both cases)

The sum of all of these fields is 22 bytes. There is no _Name_ field in these packets.

jsebechlebsky commented:
Thanks a lot for the review!

You're right about EAP-MD5, that was my mistake. I've fixed it so the length of the `optional_name` field is calculated as 
```
StrLenField("optional_name", '', length_from=lambda p: p.len - p.value_size - 6)
```
and removed change in `regression.uts`.

plorinquer commented:
You're welcome, thanks for your contribution !
